### ISSUE TITLE
Performance gains from batching on GPU

### ISSUE THREAD

**Issue by cchadowitz-pf** [davisking/dlib#1464]
* **Version**: 19.15
* **Where did you get dlib**: dlib.net
* **Platform**: Ubuntu 16.04.5 LTS, Ubuntu 16.04.4 LTS, CentOS Linux release 7.5.1804 (Core), all running a docker container based on Ubuntu 16.04.4 LTS
* **Compiler**: cmake version 3.5.1
```
$ g++ -v
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/5/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 5.4.0-6ubuntu1~16.04.10' --with-bugurl=file:///usr/share/doc/gcc-5/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-5 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-5-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-5-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-5-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10)
```
* CUDA 8.0

Related to #1159 I believe.

I was looking into how much a performance increase I could expect by batching images for inference when using the `mmod_human_face_detector.dat` face detection model on GPU.

Based loosely on the [example](http://dlib.net/dnn_mmod_face_detection_ex.cpp.html), I put together a simple app that will time the inference for a variety of batch sizes and tested this with batch sizes 1, 5, and 20 (others as well, but this was enough to show a trend) on a few different machines/hardware configurations. 

From the comment in that same example, it implies that there should be a large performance increase when batching
```
       // Note that you can process a bunch of images in a std::vector at once and it runs
        // much faster, since this will form mini-batches of images and therefore get
        // better parallelism out of your GPU hardware.  However, all the images must be
        // the same size.  To avoid this requirement on images being the same size we
        // process them individually in this example.
```

I used a docker container based on `nvidia/cuda:8.0-cudnn7-devel-ubuntu16.04` to keep the environment as similar as possible. The source code is available [here](https://gist.github.com/cchadowitz-pf/ee223f4622db01880b2e4961c73e42f7).

Some timing results I found (100 images total):

Machine | Card | Batch size 1 | Batch size 20
---------- | ---------- | ---------- | ----------
Machine 1 | GTX 1080 Ti - Pascal | ~8s | ~5s
Machine 2 | [Geforce GTX Titan X - Maxwell](https://www.geforce.com/hardware/desktop-gpus/geforce-gtx-titan-x/specifications) | ~4.5s | ~4s
Machine 2 | GTX 1080 Ti - Pascal | ~3s | ~4.1s
Machine 3 | GTX 1070 - Pascal | ~8.3s | ~8-10s

I can get more specific times for specific batch sizes on each of these cards if that would be helpful, but I was hoping to get some feedback as to whether these are somewhat expected. The last machine is by far the most puzzling.

@davisking if you have a few minutes to grab the docker image and run it on your machine, would you mind posting some times for various batch sizes? And if there's anything suspicious in my tests or procedure, please let me know.

Thank you!

Download the docker image: `docker pull cchadowitz/dlib-batch-test`

To run the app inside the container:
```
docker run --runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=0 -it --rm /bin/bash
# Inside the running container
cd ~
./dnn_mmod_face_detection_ex mmod_human_face_detector.dat 1 _frames/* # For batch size 1
./dnn_mmod_face_detection_ex mmod_human_face_detector.dat 20 _frames/* # For batch size 20
```

Attached example image
![sample_000011](https://user-images.githubusercontent.com/24324908/44871893-2148ed00-ac62-11e8-9f19-b13340a0a04e.jpg)

**Comment by davisking** [comment]
You didn't say what size of image you are using.  That is the most important statistic.  Without knowing that the timing numbers are meaningless.  For instance, I just ran an image that is 751x1002 pixels and it takes 366ms on a 1080ti.  I don't know how big your image is but it must be a monster for it to take 8 seconds to process.

Moreover, when I run a batch of 20 of these 751x1002 images the entire batch is processed in 834ms, a considerable increase in images processed per second compared to the one image at a time speed of 366ms.

It is likely that the GPU hardware is already fully saturated by your super huge image.  The point of using larger batches is just to give the GPU more data at a time.  But if you pass in a billion pixel image then one image per batch is sufficient to saturate the hardware.  Although the numbers you posted aren't really self consistent so you probably have other issues going on.  In any case, you need to be more careful when you benchmark things and pay attention to details.

**Comment by cchadowitz-pf** [comment]
So the image I attached is one of the images I used to test, the rest are included in the docker container under `_frames/`. They are all 640x360, and the code upscales them once with `pyramid_up()`. So, I don't think the GPU is saturated by far (maybe only when the batch is >30).

I've updated my app and docker image to include one forward pass before timing starts etc to warm up the cuda runtime. I've also added an additional, larger set of images as an additional set of data points. The 100 images set is in `100_frames` and the 500 images set is in `500_frames`. Both are sets of images of the same resolution, 1920x1080, and I removed any upscaling that was being done previously. So a single image will be ~2 million pixels, and a batch of 20 images will be ~40 million pixels.

Machine | Card | 100 images - batch size 1 | 100 images - batch size 20 | 500 images - batch size 1 | 500 images - batch size 20
-- | -- | -- | -- | -- | --
Machine 1 | GTX 1080 Ti - Pascal | 6.2s | 8.55s | 31.47s | 28.52s
Machine 2 | Geforce GTX Titan X - Maxwell | 6.91s | 8.28s | 36.25s | 31.73s
Machine 2 | GTX 1080 Ti - Pascal | 5.46s | 7.51s | 27.54s | 26.73s
Machine 3 | GTX 1070 - Pascal | 8.62s | 12.04s | 42.58s | 38.61s

Converted to per-image times

Machine | Card | 100 images - batch size 1 (per image) | 100 images - batch size 20 | 500 images - batch size 1 | 500 images - batch size 20
-- | -- | -- | -- | -- | --
Machine 1 | GTX 1080 Ti - Pascal | 62ms | 85.5ms | 62.94ms | 57.04ms
Machine 2 | Geforce GTX Titan X - Maxwell | 69.1ms | 82.8ms | 72.5ms | 63.46ms
Machine 2 | GTX 1080 Ti - Pascal | 54.6ms | 75.1ms | 55.08ms | 53.46ms
Machine 3 | GTX 1070 - Pascal | 86.2ms | 120.4ms | 85.16ms | 77.22ms

I see that when performing a forward pass, you can optionally specify the batch size in the forward pass call as a second parameter - how does this affect the batch sizes compared to simply passing a vector of images on its own? It seems to imply dlib will automatically create its own mini-batches from the vector provided.

Regardless, the times demonstrated above don't seem to reflect anything close to what you're seeing on your machine. Are you able to pull the docker container image I provided and run a few tests on your end as well to compare run times?

Updated source code: https://gist.github.com/cchadowitz-pf/ee223f4622db01880b2e4961c73e42f7

[Docker container image](https://hub.docker.com/r/cchadowitz/dlib-batch-test/): `docker pull cchadowitz/dlib-batch-test`

**Comment by davisking** [comment]
When you change the batch size cuda resources are reallocated, that takes a long time.  Don't time that part either.  

Also, I'm not a huge fan of docker.

**Comment by cchadowitz-pf** [comment]
I see. I've tweaked my code again to use the same batch size for the warmup stage as the rest of the stage, so that should no longer be a factor.

Updated timing results:


Machine | Card | 100 images - batch size 1 | 100 images - batch size 20 | 500 images - batch size 1 | 500 images - batch size 20
-- | -- | -- | -- | -- | --
Machine 1 | GTX 1080 Ti - Pascal | 6.56s | 4.83s | 31.17s | 27.78s
Machine 2 | Geforce GTX Titan X - Maxwell | 6.99s | 5.59s | 35.82s | 27.38s
Machine 2 | GTX 1080 Ti - Pascal | 5.58s | 4.53s | 28.08s | 23.82s
Machine 3 | GTX 1070 - Pascal | 8.43s | 7.09s | 42.33s | 35.12s

Converted to per-image times

Machine | Card | 100 images - batch size 1 (per image) | 100 images - batch size 20 | 500 images - batch size 1 | 500 images - batch size 20
-- | -- | -- | -- | -- | --
Machine 1 | GTX 1080 Ti - Pascal | 65.6ms | 48.3ms | 62.34ms | 55.56ms
Machine 2 | Geforce GTX Titan X - Maxwell | 69.9ms | 55.9ms | 71.64ms | 54.76ms
Machine 2 | GTX 1080 Ti - Pascal | 55.8ms | 45.3ms | 56.16ms | 47.64ms
Machine 3 | GTX 1070 - Pascal | 84.3ms | 70.9ms | 86.66ms | 70.24ms


The docker image was simply to make it easy to run the same tests in a fixed environment configuration.

It does seem like the varying batch sizes had a noticeable impact on performance. If I understand correctly, this is because by design dlib is having cuda allocate/reallocate only the memory necessary, and does not allocate a larger block of memory to use for entire execution up front? I know some other frameworks like Caffe and Tensorflow allocate larger blocks of memory, but that has its own issues.

Thank you for taking the time to work through this with me!
